src/ws/cockpit-session-client-certificate.c

```diff
@@ -43,65 +43,50 @@
  * desirable: Let's not get DoSed by huge certs */
 #define MAX_PEER_CERT_SIZE 100000
 
-#define CGROUP_REGEX         "^(0:|1:name=systemd):/system.slice/system-cockpithttps.slice/" \
-                             "cockpit-wsinstance-https@([0-9a-f]{64}).service$"
-#define CGROUP_REGEX_FLAGS   (REG_EXTENDED | REG_NEWLINE)
-#define CGROUP_REGEX_GROUPS  3   /* number of groups, including the complete match */
-#define CGROUP_REGEX_MATCH   2   /* the group which contains the instance */
-
-/* get our cgroup, map it to a systemd unit instance name
- * looks like 0::/system.slice/system-cockpit\x2dwsinstance\x2dhttps.slice/cockpit-wsinstance-https@123abc.service
- * returns "123abc" instance name (static string)
+/* Reads the cgroupsv2-style /proc/[pid]/cgroup file of the process,
+ * including "0://" prefix and newline.
+ *
+ * In case of cgroupsv1, look for the name=systemd controller, and fake
+ * it.
  */
-static const char *
-get_ws_https_instance (void)
+static char *
+read_proc_self_cgroup (size_t *out_length)
 {
-  int r;
-  static char buf[1024];
-  regmatch_t pmatch[CGROUP_REGEX_GROUPS];
-  regex_t preg;
-  int fd;
-
-  /* read /proc/self/cgroup */
-  fd = open ("/proc/self/cgroup", O_RDONLY | O_CLOEXEC | O_NOFOLLOW);
-  if (fd < 0)
-    {
-      warn ("Failed to open /proc/self/cgroup");
-      return NULL;
-    }
-
-  do
-    r = read (fd, buf, sizeof buf);
-  while (r < 0 && errno == EINTR);
-  if (r < 0)
-    {
-      warn ("Failed to read /proc/self/cgroup");
-      close (fd);
-      return NULL;
-    }
-  close (fd);
-  if (r == 0 || r >= sizeof buf)
-    {
-      warnx ("Read invalid size %i from /proc/self/cgroup", r);
-      return NULL;
-    }
-  buf[r] = '\0';
+  FILE *fp = fopen ("/proc/self/cgroup", "r");
 
-  /* extract the instance name */
-  r = regcomp (&preg, CGROUP_REGEX, CGROUP_REGEX_FLAGS);
-  assert (r == 0);
+  if (fp == NULL)
+    return NULL;
 
-  r = regexec (&preg, buf, CGROUP_REGEX_GROUPS, pmatch, 0);
-  regfree (&preg);
-  if (r != 0)
+  /* Support cgroups v1 by looping.
+   * Once we no longer need this support, we can drop the loop, switch
+   * to fread(), and just return the entire content of the file.
+   */
+  char buffer[1024];
+  char *result = NULL;
+  while (fgets (buffer, sizeof buffer, fp))
     {
-      warnx ("Not running in a template cgroup, unable to parse systemd unit instance.\n\n/proc/self/cgroups content follows:\n%s\n", buf);
-      return NULL;
+      assert (strchr (buffer, '\n')); /* ...or else, kernel bug. */
```

Comment:
It could catch two cases that we should never see:
 1. very long lines (ie: couldn't fit the entire line into the buffer, so we got a partial read)
 2. no newline at end of file

I don't expect to see either of those cases here, but the newline is important to the security properties here: we use it as the terminator between the cgroup header and the certificate in the file.  It's important we make sure to have it.

I think I might make this a more explicit check at the end, just before we return, where we calculate the `strlen()` anyway, that `strlen() - 1` is the `'\n'`.